src/rpc.rs

```diff
@@ -0,0 +1,177 @@
+                return result;
+            }
+            Err(_) => (),
+        }
+        let data = self.0.invalid_params();
+        Err(Error {
+            code: ErrorCode::InvalidParams,
+            message: format!("Invalid params: {:?}", value),
+            data: data.as_ref().map(to_value),
+        })
+    }
+}
+
+pub fn read_message<R>(mut reader: R) -> Result<Option<String>, Box<StdError>>
+    where R: BufRead + Read,
+{
+    let mut header = String::new();
+    let n = try!(reader.read_line(&mut header));
+    if n == 0 {
+        // EOF
+        return Ok(None);
+    }
+    if header.starts_with("Content-Length: ") {
+        let content_length = {
+            let len = header["Content-Length:".len()..].trim();
+            debug!("{}", len);
+            try!(len.parse::<usize>())
+        };
+        while header != "\r\n" {
+            header.clear();
+            try!(reader.read_line(&mut header));
+        }
+        let mut content = vec![0; content_length];
+        try!(reader.read_exact(&mut content));
+        Ok(Some(try!(String::from_utf8(content))))
+    } else {
+        Err(format!("Invalid message: `{}`", header).into())
+    }
+}
+
+pub fn write_message<W, T>(output: W, value: &T) -> io::Result<()>
+    where W: Write,
+          T: serde::Serialize,
+{
+    let response = to_string(&value).unwrap();
+    write_message_str(output, &response)
+}
+
+pub fn write_message_str<W>(mut output: W, response: &str) -> io::Result<()>
+    where W: Write,
+{
+    debug!("Respond: {}", response);
+    try!(write!(output,
+                "Content-Length: {}\r\n\r\n{}",
+                response.len(),
+                response));
+    try!(output.flush());
+    Ok(())
+}
+
+pub fn main_loop<R, W, P, F, G>(mut input: R,
+                                mut output: W,
+                                io: &IoHandler,
+                                exit_token: Arc<AtomicBool>,
+                                mut post_request_action: P,
+                                mut map_request: F,
+                                mut map_response: G)
+                                -> Result<(), Box<StdError>>
+    where R: BufRead,
+          W: Write,
+          P: FnMut() -> Result<(), Box<StdError>>,
+          F: FnMut(String) -> Result<String, Box<StdError>>,
+          G: FnMut(String) -> Result<String, Box<StdError>>,
+{
+    while !exit_token.load(atomic::Ordering::SeqCst) {
+        match try!(read_message(&mut input)) {
+            Some(json) => {
+                let json = try!(map_request(json));
+                debug!("Handle: {}", json);
+                if let Some(response) = io.handle_request_sync(&json) {
+                    let response = try!(map_response(response));
+                    try!(write_message_str(&mut output, &response));
+                    try!(output.flush());
+                }
+                post_request_action()?;
+            }
+            None => return Ok(()),
+        }
+    }
```

Comment:
Wonder if we could do a more elm-style update loop? ie. pushing the effects up higher, rather than passing in callbacks... dunno if that makes sense :/